### PR TITLE
WebServiceTarget - FlushAsync - Avoid premature flush

### LINKED ISSUE #259: Lost log statements target on application (AppDomain) close (when async writing)
Since the WebService target is asynchronous messages are lost if the application close before log statements are sent.

Here is a basic example of our program and config:

Program.cs

``` C#
using NLog;

namespace ConsoleTest
{
    internal class Program
    {
        private static void Main()
        {
            var logger = LogManager.GetLogger("MyLog");
            logger.Info("Hello, world!");
            logger.Info("Hello, world again!");
        }
    }
}
```

nlog.config

``` XML
<?xml version="1.0" encoding="utf-8" ?>
<nlog>
  <targets>
      <target name="ws" xsi:type="WebService" namespace="http://tempuri.org" methodName="SaveLog" url="http://ourserver.com/">
        <parameter name="message" type="String" layout="${message}"/>
      </target>
  </targets>

  <rules>
    <logger name="MyLog" minlevel="Info" writeTo="ws" />
  </rules>
</nlog>
```

We have tried some different approaches so far, such as wrapping the target _ws_ in `AutoFlushWrapper` (writeTo that wrapper) and even calling `LogManager.Flush()` but this did not help.  
If we use `AsyncWrapper` we seem to be able to flush out one of the messages, but not all.

### PR DESCRIPTION
**Pull request by snakefoot**
Keeps track of pendingRequests. Tries to fix #259. Instead of expecting async-target to almost handle it #1853

First idea was to keep track of each individual LogEvent (in a dictionary), but the logic (and the allocations) became a little heavy. Now it just keeps track of a pending request count, so it is not completely correct when races occurs, but I think it is good enough.

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1868)

### CODE REVIEW

File: tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs [hunk 1]
```diff
@@ -454,6 +454,15 @@ public void WebserviceTest_restapi_httppost_checkingLost()
...
 
+                // Make triple-flush to fully exercise the async flushing logic
+                try
+                {
+                    LogManager.Flush(0);
+                }
+                catch (NLog.NLogRuntimeException)
+                { }
+                LogManager.Flush(); // Waits for flush (Scheduled on top of the previous flush)
+                LogManager.Flush(); // Nothing to flush
```
Comment (304NotModified) [inline tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs]: should users also do a triple-flush?

File: src/NLog/Targets/WebServiceTarget.cs [hunk 2]
```diff
@@ -180,6 +180,7 @@ public WebServiceTarget(string name) : this()
         /// <docgen category='Web Service Options' order='10' />
         public string XmlRootNamespace { get; set; }
 
+        private readonly Dictionary<HttpWebRequest, KeyValuePair<DateTime,AsyncContinuation>> _pendingRequests = new Dictionary<HttpWebRequest, KeyValuePair<DateTime,AsyncContinuation>>();
```
Comment (304NotModified) [inline src/NLog/Targets/WebServiceTarget.cs]: isn't `HttpWebRequest` an expensive key value?

Comment (304NotModified) [inline src/NLog/Targets/WebServiceTarget.cs]: Another way would be using a linkedlist and send the info with the `AsyncContinuation` (if possible)?

File: tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs (hunk 1 above)
Comment (snakefoot) [inline tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs]: No triple flush is only for code-coverage :)

File: src/NLog/Targets/WebServiceTarget.cs [hunk 3]
```diff
@@ -226,74 +227,230 @@ protected override void DoInvoke(object[] parameters, AsyncContinuation continua
...
+                    };
+
+                DateTime utcNowTimeout = DateTime.UtcNow.AddSeconds(-120);    // Default HttpWebRequest.Timeout = 100 secs
+
+                bool foundOldWebRequest = false;
+
+                foreach (var key in new List<HttpWebRequest>(_pendingRequests.Keys))
+                {
+                    KeyValuePair<DateTime, AsyncContinuation> flushCallback;
+                    if (_pendingRequests.TryGetValue(key, out flushCallback))
```
Comment (snakefoot) [inline src/NLog/Targets/WebServiceTarget.cs]: FlushAsync is already protected by this.SyncRoot. Same lock-object is used for the other location.

**Comment by 304NotModified** [comment]
> I was hoping it was acting like a normal class, and just used default class-reference-hashcode.



yes 're right! 


 >  // The default implementation returns the sync block index for this instance.
    // Calling it on the same object multiple times will return the same value, so
    // it will technically meet the needs of a hash function, but it's less than ideal.
    // Objects (& especially value classes) should override this method.

https://referencesource.microsoft.com/#mscorlib/system/object.cs,4de9cf234d0d8b16

File: tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs (hunk 1 above)
Comment (304NotModified) [inline tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs]: sounds we should have split those tests.

File: src/NLog/Targets/WebServiceTarget.cs (hunk 3 above)
Comment (304NotModified) [inline src/NLog/Targets/WebServiceTarget.cs]: Missed that, thx!

**Comment by snakefoot** [comment]
Have now also added AutoFlushTargetWrapper into the mix, since #259 also complains that using the AutoFlushTargetWrapper didn't solve their problem.

Now I also have the crutch (AutoFlushTargetWrapper) that people can use when having a target without a proper flush-method.

File: tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs (hunk 1 above)
Comment (snakefoot) [inline tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs]: Done

**Comment by 304NotModified** [comment]
Reviewed 15 of 17 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs, line 626 at r3](https://reviewable.io:443/reviews/nlog/nlog/1868#-KaEl1TuhuaxsuL6eXyS:-KaEl1TuhuaxsuL6eXyT:b-st8lp5) ([raw file](https://github.com/nlog/nlog/blob/89a85193b1a162bb4437493085f040e445a71222/tests/NLog.UnitTests/Targets/WebServiceTargetTests.cs#L626)):*
> ```Smalltalk
>                     CountdownEvent = new CountdownEvent(expectedMessages);
>                 else
>                     CountdownEvent = null;
> ```

nice one :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nlog/nlog/1868)*

**Comment by snakefoot** [comment]
@304NotModified  Think I have broken something, as the test-builds are now failing quite frequently.

**Pull request merged**